File: src/Conjecture.NUnit.Tests/ReproExportIntegrationTests.cs
// Copyright (c) 2026 Kim Ommundsen. Licensed under the MPL-2.0.
// See LICENSE.txt in the project root or https://mozilla.org/MPL/2.0/

using System.Reflection;

using Conjecture.Core;
using Conjecture.Core.Internal;

using NUnit.Framework;

using ConjecturePropertyAttribute = Conjecture.NUnit.PropertyAttribute;

namespace Conjecture.NUnit.Tests;

[TestFixture]
public class ReproExportIntegrationTests
{
#pragma warning disable IDE0060
    private static void IntProperty(int x) { }
#pragma warning restore IDE0060

    private static ParameterInfo[] Params(string methodName)
    {
        return typeof(ReproExportIntegrationTests)
            .GetMethod(methodName, BindingFlags.NonPublic | BindingFlags.Static)!
            .GetParameters();
    }

    [Test]
    public void PropertyAttribute_ExportReproductionOnFailure_DefaultIsFalse()
    {
        ConjecturePropertyAttribute attr = new();
        Assert.That(attr.ExportReproductionOnFailure, Is.False);
    }

    [Test]
    public void PropertyAttribute_ReproductionOutputPath_DefaultIsConjecturePath()
    {
        ConjecturePropertyAttribute attr = new();
        Assert.That(attr.ReproductionOutputPath, Is.EqualTo(".conjecture/repros/"));
    }

    [Test]
    public async Task ExportReproductionOnFailureTrue_FailingProperty_WritesCsFile()
    {
        string tempDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
        Directory.CreateDirectory(tempDir);
        try
        {
            ConjectureSettings settings = new()
            {
                MaxExamples = 50,
                Seed = 1UL,
                ExportReproductionOnFailure = true,
                ReproductionOutputPath = tempDir,
            };
            ParameterInfo[] parameters = Params(nameof(IntProperty));

            TestRunResult result = await TestRunner.Run(settings, data =>
            {
                object[] args = SharedParameterStrategyResolver.Resolve(parameters, data);
                if ((int)args[0] > 5) { throw new Exception("fail"); }
            });

            Assert.That(result.Passed, Is.False);

            ConjectureData replay = ConjectureData.ForRecord(result.Counterexample!);
            object[] shrunkArgs = SharedParameterStrategyResolver.Resolve(parameters, replay);
            IEnumerable<(string Name, object? Value, Type Type)> paramTuples =
                parameters.Select((p, i) => (p.Name!, (object?)shrunkArgs[i], p.ParameterType));
            ReproContext context = new(
                nameof(ReproExportIntegrationTests),
                nameof(IntProperty),
                false,
                paramTuples,
                result.Seed!.Value,
                result.ExampleCount,
                result.ShrinkCount,
                TestFramework.NUnit,
                DateTimeOffset.UtcNow);

            ReproFileBuilder.WriteToFile(context, settings.ReproductionOutputPath);
            string[] files = Directory.GetFiles(tempDir, "*.cs");
            Assert.That(files, Is.Not.Empty);
        }
        finally
        {
            Directory.Delete(tempDir, true);
        }
    }
}

File: src/Conjecture.FSharp.Expecto.Tests/ReproExportIntegrationTests.fs
// Copyright (c) 2026 Kim Ommundsen. Licensed under the MPL-2.0.
// See LICENSE.txt in the project root or https://mozilla.org/MPL/2.0/

module Conjecture.FSharp.Expecto.Tests.ReproExportIntegrationTests

open System
open System.IO
open Expecto
open Conjecture
open Conjecture.FSharp.Expecto

let private runTest (t: Test) : int =
    runTestsWithCLIArgs [] [||] t

let tests =
    testList "Conjecture.FSharp.Expecto.repro-export" [

        testCase "propertyWithRepro writes a .cs file when ExportOnFailure is true and property fails" <| fun () ->
            let tempDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName())
            Directory.CreateDirectory(tempDir) |> ignore
            try
                let repro = { ExportOnFailure = true; OutputPath = tempDir }
                let t = propertyWithRepro "fails" (fun (x: int) -> x < -1_000_000_000) repro
                let _ = try runTest t with _ -> 1
                let files = Directory.GetFiles(tempDir, "*.cs")
                Expect.isNonEmpty files "expected at least one .cs repro file"
            finally
                try Directory.Delete(tempDir, true) with _ -> ()

        testCase "propertyWithRepro writes no file when ExportOnFailure is false" <| fun () ->
            let tempDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName())
            Directory.CreateDirectory(tempDir) |> ignore
            try
                let repro = { ExportOnFailure = false; OutputPath = tempDir }
                let t = propertyWithRepro "fails" (fun (x: int) -> x < -1_000_000_000) repro
                let _ = try runTest t with _ -> 1
                let files = Directory.GetFiles(tempDir, "*.cs")
                Expect.isEmpty files "expected no repro file when export disabled"
            finally
                try Directory.Delete(tempDir, true) with _ -> ()
    ]
